core/src/main/java/org/apache/iceberg/io/datafile/DataFileServiceRegistry.java

```diff
@@ -0,0 +1,420 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io.datafile;
+
+import java.util.Map;
+import java.util.ServiceLoader;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Objects;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Registry which maintains the available {@link ReaderService} and {@link WriterService}
+ * implementations. Based on the `file format`, the required `data type` and the reader/writer
+ * `builderType` the registry returns the correct reader and writer service implementations. These
+ * services could be used to generate the correct reader and writer builders.
+ */
+public final class DataFileServiceRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(DataFileServiceRegistry.class);
+  private static final Map<Key, ReaderService> READ_BUILDERS = Maps.newConcurrentMap();
+  private static final Map<Key, WriterService<?>> WRITE_BUILDERS = Maps.newConcurrentMap();
+
+  static {
+    for (ReaderService service : ServiceLoader.load(ReaderService.class)) {
+      if (READ_BUILDERS.containsKey(service.key())) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Read service %s clashes with %s. Both serves %s",
+                service.getClass(), READ_BUILDERS.get(service.key()), service.key()));
+      }
+
+      READ_BUILDERS.putIfAbsent(service.key(), service);
+    }
+
+    for (WriterService<?> service : ServiceLoader.load(WriterService.class)) {
+      if (WRITE_BUILDERS.containsKey(service.key())) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Write service %s clashes with %s. Both serves %s",
+                service.getClass(), WRITE_BUILDERS.get(service.key()), service.key()));
+      }
+
+      WRITE_BUILDERS.putIfAbsent(service.key(), service);
+    }
+
+    LOG.info("DataFileServices found: readers={}, writers={}", READ_BUILDERS, WRITE_BUILDERS);
+  }
+
+  private DataFileServiceRegistry() {}
+
+  /**
+   * Provides a reader for the given input file which returns objects with a given returnType.
+   *
+   * @param format of the file to read
+   * @param returnType returned by the reader
+   * @param inputFile to read
+   * @param readSchema to use when reading the data file
+   * @return {@link ReaderBuilder} for building the actual reader
+   */
+  public static ReaderBuilder read(
```

Comment:
Nomenclature: maybe `newReaderBuilder` or just `readerBuilder` instead of `read`?

Reply:
Done